File: Realization_GO/queue.go
package main

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"os"
	"strconv"
)

// Queue - очередь на кольцевом буфере
type Queue struct {
	arr         []int
	capacity    int
	frontIdx    int
	rearIdx     int
	currentSize int
}

// NewQueue - конструктор
func NewQueue() *Queue {
	q := &Queue{
		arr:         make([]int, 4),
		capacity:    4,
		frontIdx:    0,
		rearIdx:     0,
		currentSize: 0,
	}
	return q
}

// resize - изменение размера
func (q *Queue) resize() {
	newCapacity := q.capacity * 2
	newArr := make([]int, newCapacity)

	i := 0
	j := q.frontIdx
	for i < q.currentSize {
		newArr[i] = q.arr[j]
		j = (j + 1) % q.capacity
		i++
	}

	q.arr = newArr
	q.capacity = newCapacity
	q.frontIdx = 0
	q.rearIdx = q.currentSize
}

// Push - добавление элемента
func (q *Queue) Push(value int) {
	if q.currentSize == q.capacity {
		q.resize()
	}

	q.arr[q.rearIdx] = value
	q.rearIdx = (q.rearIdx + 1) % q.capacity
	q.currentSize++

	fmt.Printf("Добавлен элемент: %d\n", value)
}

// Pop - удаление элемента
func (q *Queue) Pop() {
	if q.IsEmpty() {
		fmt.Println("Очередь пуста! Нельзя удалить элемент.")
		return
	}

	frontValue := q.arr[q.frontIdx]
	q.frontIdx = (q.frontIdx + 1) % q.capacity
	q.currentSize--

	fmt.Printf("Удален элемент: %d\n", frontValue)
}

// Front - получение первого элемента
func (q *Queue) Front() int {
	if q.IsEmpty() {
		fmt.Println("Очередь пуста!")
		return -1
	}
	return q.arr[q.frontIdx]
}

// IsEmpty - проверка на пустоту
func (q *Queue) IsEmpty() bool {
	return q.currentSize == 0
}

// Size - получение размера
func (q *Queue) Size() int {
	return q.currentSize
}

// Display - отображение очереди
func (q *Queue) Display() {
	if q.IsEmpty() {
		fmt.Println("Очередь пуста!")
		return
	}

	fmt.Print("Содержимое очереди: ")
	idx := q.frontIdx
	for i := 0; i < q.currentSize; i++ {
		fmt.Printf("%d ", q.arr[idx])
		idx = (idx + 1) % q.capacity
	}
	fmt.Println()
}

// Clear - очистка очереди
func (q *Queue) Clear() {
	q.frontIdx = 0
	q.rearIdx = 0
	q.currentSize = 0
	fmt.Println("Очередь очищена!")
}

// SaveToFile - сохранение в текстовый файл
func (q *Queue) SaveToFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия файла: %s\n", filename)
		return err
	}
	defer file.Close()

	idx := q.frontIdx
	for i := 0; i < q.currentSize; i++ {
		_, err := fmt.Fprintf(file, "%d ", q.arr[idx])
		if err != nil {
			return err
		}
		idx = (idx + 1) % q.capacity
	}

	fmt.Printf("Очередь сохранена в файл: %s\n", filename)
	return nil
}

// LoadFromFile - загрузка из текстового файла
func (q *Queue) LoadFromFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия файла: %s\n", filename)
		return err
	}
	defer file.Close()

	q.Clear()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanWords)

	for scanner.Scan() {
		value, err := strconv.Atoi(scanner.Text())
		if err == nil {
			q.Push(value)
		}
	}

	fmt.Printf("Очередь загружена из файла: %s\n", filename)
	return scanner.Err()
}

// SaveToBinaryFile - сохранение в бинарный файл
func (q *Queue) SaveToBinaryFile(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия бинарного файла для записи: %s\n", filename)
		return err
	}
	defer file.Close()

	queueSize := int64(q.currentSize)
	err = binary.Write(file, binary.LittleEndian, queueSize)
	if err != nil {
		return err
	}

	idx := q.frontIdx
	for i := 0; i < q.currentSize; i++ {
		err = binary.Write(file, binary.LittleEndian, int32(q.arr[idx]))
		if err != nil {
			return err
		}
		idx = (idx + 1) % q.capacity
	}

	fmt.Printf("Очередь сохранена в бинарный файл: %s\n", filename)
	return nil
}

// LoadFromBinaryFile - загрузка из бинарного файла
func (q *Queue) LoadFromBinaryFile(filename string) error {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Ошибка открытия бинарного файла для чтения: %s\n", filename)
		return err
	}
	defer file.Close()

	q.Clear()

	var queueSize int64
	err = binary.Read(file, binary.LittleEndian, &queueSize)
	if err != nil {
		return err
	}

	for i := 0; i < int(queueSize); i++ {
		var value int32
		err = binary.Read(file, binary.LittleEndian, &value)
		if err != nil {
			return err
		}
		q.Push(int(value))
	}

	fmt.Printf("Очередь загружена из бинарного файла: %s\n", filename)
	return nil
}

File: LR3/Realization_C++/Queue.cpp
#include "Queue.h"
#include <iostream>
#include <fstream>

using namespace std;

Queue::Queue() : arr(nullptr), capacity(0), front_idx(0), rear_idx(0), current_size(0) {
    resize();
}

Queue::~Queue() {
    delete[] arr;
}

Queue::Queue(const Queue& other) : arr(nullptr), capacity(0), front_idx(0), rear_idx(0), current_size(0) {
    *this = other;
}

Queue& Queue::operator=(const Queue& other) {
    if (this != &other) {
        delete[] arr;
        capacity = other.capacity;
        current_size = other.current_size;
        front_idx = 0;
        rear_idx = 0;
        
        arr = new int[capacity];
        size_t i = 0;
        size_t j = other.front_idx;
        while (i < current_size) {
            arr[i] = other.arr[j];
            j = (j + 1) % capacity;
            i++;
        }
        rear_idx = current_size;
    }
    return *this;
}

void Queue::resize() {
    size_t new_capacity = capacity == 0 ? 4 : capacity * 2;
    int* new_arr = new int[new_capacity];
    
    size_t i = 0;
    size_t j = front_idx;
    while (i < current_size) {
        new_arr[i] = arr[j];
        j = (j + 1) % capacity;
        i++;
    }
    
    delete[] arr;
    arr = new_arr;
    capacity = new_capacity;
    front_idx = 0;
    rear_idx = current_size;
}

void Queue::push(int value) {
    if (current_size == capacity) {
        resize();
    }
    
    arr[rear_idx] = value;
    rear_idx = (rear_idx + 1) % capacity;
    current_size++;
    
    cout << "Добавлен элемент: " << value << endl;
}

void Queue::pop() {
    if (isEmpty()) {
        cout << "Очередь пуста! Нельзя удалить элемент." << endl;
        return;
    }
    
    int frontValue = arr[front_idx];
    front_idx = (front_idx + 1) % capacity;
    current_size--;
    
    cout << "Удален элемент: " << frontValue << endl;
}

int Queue::front() {
    if (isEmpty()) {
        cout << "Очередь пуста!" << endl;
        return -1;
    }
    return arr[front_idx];
}

bool Queue::isEmpty() {
    return current_size == 0;
}

size_t Queue::size() {
    return current_size;
}

void Queue::display() {
    if (isEmpty()) {
        cout << "Очередь пуста!" << endl;
        return;
    }

    cout << "Содержимое очереди: ";
    size_t idx = front_idx;
    for (size_t i = 0; i < current_size; i++) {
        cout << arr[idx] << " ";
        idx = (idx + 1) % capacity;
    }
    cout << endl;
}

void Queue::clear() {
    front_idx = 0;
    rear_idx = 0;
    current_size = 0;
    cout << "Очередь очищена!" << endl;
}

void Queue::saveToFile(const string& filename) {
    ofstream file(filename);
    if (file.is_open()) {
        size_t idx = front_idx;
        for (size_t i = 0; i < current_size; i++) {
            file << arr[idx] << " ";
            idx = (idx + 1) % capacity;
        }
        file.close();
        cout << "Очередь сохранена в файл: " << filename << endl;
    } else {
        cout << "Ошибка открытия файла: " << filename << endl;
    }
}

void Queue::loadFromFile(const string& filename) {
    ifstream file(filename);
    if (file.is_open()) {
        clear();
        int value;
        while (file >> value) {
            push(value);
        }
        file.close();
        cout << "Очередь загружена из файла: " << filename << endl;
    } else {
        cout << "Ошибка открытия файла: " << filename << endl;
    }
}

// Бинарная сериализация
void Queue::saveToBinaryFile(const string& filename) {
    ofstream file(filename, ios::binary);
    if(file.is_open()) {
        size_t queueSize = current_size;
        file.write(reinterpret_cast<const char*>(&queueSize), sizeof(size_t));
        
        size_t idx = front_idx;
        for(size_t i = 0; i < current_size; i++) {
            file.write(reinterpret_cast<const char*>(&arr[idx]), sizeof(int));
            idx = (idx + 1) % capacity;
        }
        
        file.close();
        cout << "Очередь сохранена в бинарный файл: " << filename << endl;
    } else {
        cout << "Ошибка открытия бинарного файла для записи: " << filename << endl;
    }
}

void Queue::loadFromBinaryFile(const string& filename) {
    ifstream file(filename, ios::binary);
    if(file.is_open()) {
        clear();
        
        size_t queueSize;
        file.read(reinterpret_cast<char*>(&queueSize), sizeof(size_t));
        
        for(size_t i = 0; i < queueSize; i++) {
            int value;
            file.read(reinterpret_cast<char*>(&value), sizeof(int));
            push(value);
        }
        
        file.close();
        cout << "Очередь загружена из бинарного файла: " << filename << endl;
    } else {
        cout << "Ошибка открытия бинарного файла для чтения: " << filename << endl;
    }
}